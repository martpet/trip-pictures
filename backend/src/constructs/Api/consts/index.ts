export * from './authorizationHeader';
export * from './lambdaEdgeViewerEvent';
