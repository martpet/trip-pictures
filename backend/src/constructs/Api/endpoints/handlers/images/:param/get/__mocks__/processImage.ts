export const processImage = vi
  .fn()
  .mockName('processImage')
  .mockResolvedValue('dummyBase64EncodedImage');
