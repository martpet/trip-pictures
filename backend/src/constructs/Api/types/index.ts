export * from './Cookies';
export * from './EnvVars';
export * from './LambdaEdgeOriginRequestHandler';
export * from './LambdaEdgeResponse';
export * from './OauthCookieProps';
export * from './OauthTokens';
export * from './RouteHeaders';
