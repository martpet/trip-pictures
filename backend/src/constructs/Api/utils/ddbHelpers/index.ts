export * from './ddbBatchGet';
export * from './ddbBatchWrite';
export * from './ddbScan';
