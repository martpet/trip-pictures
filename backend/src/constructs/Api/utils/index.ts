export * from './cookieName';
export * from './ddbHelpers';
export * from './edgeErrorResponse';
export * from './errorResponse';
export * from './getIdTokenPayload';
export * from './parseEventCookies';
export * from './parseLambdaEdgeEventCookies';
export * from './testHelpers';
