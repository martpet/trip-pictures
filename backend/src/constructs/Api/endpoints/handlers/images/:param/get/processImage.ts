import { SdkStreamMixin } from '@aws-sdk/types';
import { DetectModerationLabelsCommand, RekognitionClient } from 'lambda-layer';
import sharp from 'sharp';

const rekognition = new RekognitionClient({});

type Props = {
  stream: SdkStreamMixin;
  quality?: number;
  width?: number;
  height?: number;
};

export async function processImage({
  stream,
  width = 1500,
  height,
  quality = 100,
}: Props) {
  const uint8 = await stream.transformToByteArray();
  const image = sharp(uint8).rotate().resize(width, height).jpeg({ quality });
  let buffer = await image.toBuffer();

  const moderationCommand = new DetectModerationLabelsCommand({
    Image: { Bytes: buffer },
    MinConfidence: 90,
  });

  const { ModerationLabels } = await rekognition.send(moderationCommand);

  const isModerated = ModerationLabels?.some(({ Name }) => Name === 'Explicit Nudity');

  if (isModerated) {
    image.blur(75);
    buffer = await image.toBuffer();
  }

  const base64 = buffer.toString('base64');
  return base64;
}
