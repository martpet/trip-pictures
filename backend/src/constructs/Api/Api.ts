import { HttpApi } from '@aws-cdk/aws-apigatewayv2-alpha';
import { Fn, NestedStack, RemovalPolicy } from 'aws-cdk-lib';
import { Distribution } from 'aws-cdk-lib/aws-cloudfront';
import { HttpOrigin } from 'aws-cdk-lib/aws-cloudfront-origins';
import { ARecord, RecordTarget } from 'aws-cdk-lib/aws-route53';
import { CloudFrontTarget } from 'aws-cdk-lib/aws-route53-targets';
import { Bucket } from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';

import { Auth, Photos, Zone } from '~/constructs';
import { apiSubdomain, appEnvs } from '~/consts';
import { getCdkEnv } from '~/utils';

import { createDistroBehaviors } from './endpoints/createDistroBehaviors';
import { createRoutes } from './endpoints/createRoutes';

type Props = {
  zone: Zone;
  auth: Auth;
  photos: Photos;
};

export class Api extends NestedStack {
  constructor(scope: Construct, { zone, auth, photos }: Props) {
    super(scope, 'Api');

    const envName = getCdkEnv();
    const { appDomain } = appEnvs[envName];
    const apiDomain = `${apiSubdomain}.${appDomain}`;
    const api = new HttpApi(this, 'http-api');
    const origin = new HttpOrigin(Fn.select(1, Fn.split('://', api.apiEndpoint)));

    const logBucket = new Bucket(this, 'Distro-Log-Bucket', {
      removalPolicy: RemovalPolicy.DESTROY,
      autoDeleteObjects: true,
    });

    const distribution = new Distribution(this, 'Distro', {
      defaultBehavior: { origin },
      domainNames: [apiDomain],
      certificate: zone.certificate,
      enableLogging: true,
      logIncludesCookies: true,
      logBucket,
    });

    createRoutes({ scope, api, auth, photos });

    createDistroBehaviors({
      scope,
      api,
      auth,
      distribution,
      origin,
    });

    new ARecord(this, 'A-Record', {
      zone: zone.hostedZone,
      recordName: apiSubdomain,
      target: RecordTarget.fromAlias(new CloudFrontTarget(distribution)),
    });
  }
}
