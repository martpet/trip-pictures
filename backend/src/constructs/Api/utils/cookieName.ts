import { AllCookies } from '~/constructs/Api/types';

export const cookieName = (name: keyof AllCookies) => name;
