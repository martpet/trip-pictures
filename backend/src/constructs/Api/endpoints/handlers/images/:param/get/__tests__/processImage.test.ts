import {
  DetectModerationLabelsCommand,
  RekognitionClient,
} from '@aws-sdk/client-rekognition';
import { mockClient } from 'aws-sdk-client-mock';
import sharp from 'sharp';

import { itCalls, itResolves, itSendsAwsCommand } from '~/constructs/Api/utils';

import { processImage } from '../processImage';

vi.mock('sharp');

const rekognitionMock = mockClient(RekognitionClient);

beforeEach(() => {
  rekognitionMock.reset();
  rekognitionMock.on(DetectModerationLabelsCommand).resolves({});
});

const args = [
  {
    stream: {
      transformToByteArray: () => Promise.resolve('dummyByteArray'),
    },
    photoBucket: 'dummyPhotoBucketName',
    s3ObjectName: 'dummyS3ObjectName',
    width: 500,
    height: 400,
    quality: 80,
  },
] as unknown as Parameters<typeof processImage>;

describe('processImage', async () => {
  itSendsAwsCommand(DetectModerationLabelsCommand, rekognitionMock, processImage, args);
  itCalls(sharp, processImage, args);
  itCalls(sharp().rotate, processImage, args);
  itCalls(sharp().resize, processImage, args);
  itCalls(sharp().jpeg, processImage, args);
  itCalls(sharp().toBuffer, processImage, args);
  itCalls((await sharp().toBuffer()).toString, processImage, args);
  itResolves(processImage, args);

  describe.each(['width', 'height', 'quality'])('when "%s" prop is missing', (key) => {
    const propsClone = { ...args[0] };
    const argsClone = [propsClone];
    delete propsClone[key as keyof typeof propsClone];
    itCalls(sharp().jpeg, processImage, argsClone);
    itCalls(sharp().resize, processImage, argsClone);
  });

  describe('when "DetectModerationLabelsCommand" returns "Explicit Nudity"', () => {
    beforeEach(() => {
      rekognitionMock.on(DetectModerationLabelsCommand).resolves({
        ModerationLabels: [{ Name: 'Explicit Nudity' }],
      });
    });

    itCalls(sharp().blur, processImage, args);
    itCalls(sharp().toBuffer, processImage, args);
  });
});
