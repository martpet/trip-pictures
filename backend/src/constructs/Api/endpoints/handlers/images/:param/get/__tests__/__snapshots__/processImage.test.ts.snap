// Vitest Snapshot v1

exports[`processImage > calls "jpeg" with correct args 1`] = `
[
  [
    {
      "quality": 80,
    },
  ],
]
`;

exports[`processImage > calls "resize" with correct args 1`] = `
[
  [
    500,
    400,
  ],
]
`;

exports[`processImage > calls "rotate" with correct args 1`] = `
[
  [],
]
`;

exports[`processImage > calls "sharp" with correct args 1`] = `
[
  [
    "dummyByteArray",
  ],
]
`;

exports[`processImage > calls "toBuffer" with correct args 1`] = `
[
  [],
]
`;

exports[`processImage > calls "toString" with correct args 1`] = `
[
  [
    "base64",
  ],
]
`;

exports[`processImage > resolves with a correct value 1`] = `"dummyBufferToString"`;

exports[`processImage > sends "DetectModerationLabelsCommand" from "RekognitionClient" with correct args 1`] = `
{
  "Image": {
    "Bytes": {
      "toString": [MockFunction toString] {
        "calls": [
          [
            "base64",
          ],
        ],
        "results": [
          {
            "type": "return",
            "value": "dummyBufferToString",
          },
        ],
      },
    },
  },
  "MinConfidence": 90,
}
`;

exports[`processImage > when "DetectModerationLabelsCommand" returns "Explicit Nudity" > calls "blur" with correct args 1`] = `
[
  [
    75,
  ],
]
`;

exports[`processImage > when "DetectModerationLabelsCommand" returns "Explicit Nudity" > calls "toBuffer" with correct args 1`] = `
[
  [],
  [],
]
`;

exports[`processImage > when "height" prop is missing > calls "jpeg" with correct args 1`] = `
[
  [
    {
      "quality": 80,
    },
  ],
]
`;

exports[`processImage > when "height" prop is missing > calls "resize" with correct args 1`] = `
[
  [
    500,
    undefined,
  ],
]
`;

exports[`processImage > when "quality" prop is missing > calls "jpeg" with correct args 1`] = `
[
  [
    {
      "quality": 100,
    },
  ],
]
`;

exports[`processImage > when "quality" prop is missing > calls "resize" with correct args 1`] = `
[
  [
    500,
    400,
  ],
]
`;

exports[`processImage > when "width" prop is missing > calls "jpeg" with correct args 1`] = `
[
  [
    {
      "quality": 80,
    },
  ],
]
`;

exports[`processImage > when "width" prop is missing > calls "resize" with correct args 1`] = `
[
  [
    1500,
    400,
  ],
]
`;
